### DISCUSSION TITLE
Database & Failed microSD Card

### DISCUSSION THREAD

**Discussion by Cube3SGM** [David00/rpi-power-monitor#100] [Q&A]
Hello David,

Running V3.0, Circuit monitoring, Pi 3

1.   In the database I have a spike in the data. When viewing a long period, say 60 days, it shows the spike at 10s of thousands of kWh a month or so ago.  Is there a way to correct this?
2.  (Not related to above, as this one happened yesterday) I have had an outside armored cable cut on one of the circuits, straight short between live and earth!  This I think has corrupted the SD card and RPI Monitor no longer boots. I have no clue if it was the spike in current or the unceremonious shut down or something else!  The Pi attempts to boot but gives lots of errors and stops, turns the activity lights off.  I can successfully boot the Pi and run RPi Monitor with another SD Card.  Can I retrieve the database for a new install of V3.0? Oh and yeah....no back-up...sorry :)

Regards
Simon

**Comment by David00** [comment]
Hi @Cube3SGM, how are you viewing the spike in the data? In Grafana? If so, which dashboard?  If directly in the Influx shell, can you share the query?  And, by "correct this", do you mean to remove the spike?  I'm asking because it could be a visualization error.  If you can share a screenshot in your reply, that would be great too!

Regarding the boot issue, it's likely that your card has failed. This is pretty common on Pis - especially with the non-high-endurance line cards.  It might be possible to get some data off, but you'll need another microSD card (or USB drive) to boot your Pi from. 

There is some related info in #34 where @bobstanl went through two microSD card failures.  There are some steps to try getting data off the bad card scattered through that discussion, but I've consolidated the steps below so you don't have to hunt for them:

To start, you'll need a new microSD running my custom OS (the process relies on having InfluxDB already installed on the fresh image).

1. Assuming you've booted your Pi from a working card with a fresh install of my custom OS, connect your bad card to the Pi with a USB microSD card reader.
2. Run `sudo fdisk -l | grep "/dev/sd*"`. It should show something like sda1 or sdb1.  Note the sdxx that it shows because you'll need it in the following steps.
3. Create a new directory to mount the bad sdcard to: `mkdir ~/sdcard`
4. `sudo mount /dev/sdxx ~/sdcard`  (Replace sdxx with your actual sd**)

If you make it this far, the bad card is now mounted to `~/sdcard`. The influx database exists in `/opt/influxdb/`, but since that's now inside of `~/sdcard/`, it's actually at the path `~/sdcard/opt/influxdb/`.

Assuming you're using my custom OS image, which already has InfluxDB installed, stop InfluxDB with `sudo systemctl stop influxdb.service`.

Then, you can simply attempt to copy the entire `~/sdcard/opt/influxdb/` directory to your new Pi's `/opt/influxdb/` directory with:

`sudo cp -r ~/sdcard/opt/influxdb/* /opt/influxdb/`

Then, if/when that finishes successfully, you can restart Influx:

`sudo systemctl start influxdb.service`

**Reply by Cube3SGM** [reply]
Hello David,
Thank you for the prompt reply.
the output of fdisk was
sda1 8192 532479 524288 256M c W95 FAT32 (LBA)
sda2 532480 30392319 29859840 14.2G 83 Linux

I tried sudo mount /dev/sda2 ~/sdcard
returned
EXT4-fs error (device sda2): ext4_validate_block_bitmap:390 comn ext4lazyinit: bg 47: bad block bitmap checksum

The copy statement returned a similar error statement.  I assume the card is corrupted, oh well, thank you for helping me try.

With regards to the spike reading.  It was through Grafana and was on one of the kWh channels.  I can't send any more info or screen shots as this DB is now corrupt I assume.
Thanks Simon

**Reply by vulture2600** [reply]
Just an FYI for any Pi users: You can boot from a USB stick or a hard drive and not use an SD card at all. I've had too many SD cards go corrupt and I've switched all of my Pi's to USB sticks and I've never looked back.
